File: scala/Practice2/DynamicProgramming/Section/DarumaOtoshi.scala
// http://judge.u-aizu.ac.jp/onlinejudge/description.jsp?id=1611

object Main {

  implicit class Pipeline[T](x: T) {
    def |>[S](f: T => S): S = f(x)
  }

  def main(args: Array[String]) = (
    Stream.continually(io.StdIn.readLine())
      .takeWhile(_ != "0")
      .foldLeft((0, List[Int](), List[Array[Int]]())) {
        case ((i, ns, wss), s) => (
          if (i % 2 == 0) (
            s.toInt
            |> (n => (i + 1, n :: ns, wss))
          )
          else (
            s.split(" ").map(_.toInt)
            |> (ws => (i + 1, ns, ws :: wss))
          )
        )
      }
      |> { case (_x, ns, wss) => (ns zip wss) }
      |> (_.reverse)
      |> (_.map { case (n, ws) => (
        (
          { case (a, b) => ws(a) - ws(b) |> Math.abs |> (_ <= 1) }: (Int, Int) => Boolean,
          (1 to n).map(_x => Array.fill[Boolean](n)(true)).toArray,
          (1 to n).map(_x => Array.fill[Int](n)(0)).toArray
        )
        |> { case (isR, memR, memC) => (
          (
            { case (s, e) => (
              (
                (e - s) match {
                  case x if x % 2 == 0 => false
                  case 1 => isR(s, e)
                  case _ =>
                    (isR(s, e) && memR(s + 1)(e - 1)) ||
                      (memR(s)(s + 1) && memR(s + 2)(e))

                }
              )
              |> (x => ((memR(s)(e) = x) |> (_x => x)))
            )} : (Int, Int) => Boolean
          )
          |> (fnR =>
            (1 until n)
              .map((i) =>
                (for { j <- (0 until (n - i)) } yield (j, j + i))
                  .map { case (s, e) => (
                    (
                      if (fnR(s, e))
                        e - s + 1
                      else
                        (for { k <- (s until e) } yield ((s, k), (k + 1, e)))
                          .map { case ((s1, e1), (s2, e2)) =>
                            memC(s1)(e1) + memC(s2)(e2)
                          }
                          .max
                    )
                    |> (x => memC(s)(e) = x)
                  )}
              )
          )
          |> (_x => memC(0)(n - 1))
        )}
      )})
      |> (_ foreach println)
  )
}

// http://judge.u-aizu.ac.jp/onlinejudge/review.jsp?rid=5225206

File: fs/Practice2/DynamicProgramming/Section/DarumaOtoshi.fs
// http://judge.u-aizu.ac.jp/onlinejudge/description.jsp?id=1611

Seq.initInfinite (fun _ -> stdin.ReadLine ())
|> Seq.takeWhile ((<>) "0")
|> Seq.fold (
  fun (i, ns, wss) s ->
    if i % 2 = 0
    then (
      s
      |> int
      |> fun n -> (i + 1, n :: ns, wss)
    )
    else (
      s.Split(' ')
      |> Array.map int
      |> fun ws -> (i + 1, ns, ws :: wss)
    )
) (0, [], [])
|> fun (_, ns, wss) -> (List.zip ns wss)
|> List.rev
|> List.map (
  fun (n, ws) ->
    (
      (fun a b -> ws.[a] - ws.[b] |> abs |> ((>=) 1)),
      (Array.init n (fun _ -> Array.create n true)),
      (Array.init n (fun _ -> Array.create n 0))
    )
    |> fun (isR, memR, memC) ->
      (
        fun s e ->
          (
            match e - s with
            | x when x % 2 = 0 -> false
            | 1 -> isR s e
            | _ ->
              (isR s e && memR.[s + 1].[e - 1])
                || (memR.[s].[s + 1] && memR.[s + 2].[e])
          )
          |> fun x -> (memR.[s].[e] <- x) |> fun () -> x
      )
      |> fun fnR ->
        [1..(n - 1)]
        |> List.map (
          fun i ->
            [for j in 0..(n - i - 1) -> (j, j + i)]
            |> List.map (
              fun (s, e) ->
                (
                  if fnR s e
                  then (e - s + 1)
                  else
                    [for k in s..(e - 1) -> ((s, k), (k + 1, e))]
                    |> List.map (
                      fun ((s1, e1), (s2, e2)) ->
                        memC.[s1].[e1] + memC.[s2].[e2]
                    )
                    |> List.max
                )
                |> fun x -> memC.[s].[e] <- x
            )
        )
      |> fun _ -> memC.[0].[n - 1]
)
|> List.iter (printfn "%d")

// ----

// input:
// 14
// 8 7 1 4 3 5 4 1 6 8 10 4 6 5
// 0

// memC:
// [|
//   [| 0;  2;  2;  2;  4;  4;  6;  8;  8; 10; 10; 10; 10; 12 |];
//   [| 0;  0;  0;  0;  2;  2;  4;  6;  8;  8;  8;  8;  8; 10 |];
//   [| 0;  0;  0;  0;  2;  2;  4;  6;  6;  6;  6;  6;  6;  8 |];
//   [| 0;  0;  0;  0;  2;  2;  4;  4;  4;  4;  4;  4;  4;  6 |];
//   [| 0;  0;  0;  0;  0;  0;  2;  2;  2;  2;  2;  2;  2;  4 |];
//   [| 0;  0;  0;  0;  0;  0;  2;  2;  2;  2;  2;  2;  2;  4 |];
//   [| 0;  0;  0;  0;  0;  0;  0;  0;  0;  0;  0;  0;  0;  2 |];
//   [| 0;  0;  0;  0;  0;  0;  0;  0;  0;  0;  0;  0;  0;  2 |];
//   [| 0;  0;  0;  0;  0;  0;  0;  0;  0;  0;  0;  0;  0;  2 |];
//   [| 0;  0;  0;  0;  0;  0;  0;  0;  0;  0;  0;  0;  0;  2 |];
//   [| 0;  0;  0;  0;  0;  0;  0;  0;  0;  0;  0;  0;  0;  2 |];
//   [| 0;  0;  0;  0;  0;  0;  0;  0;  0;  0;  0;  0;  0;  2 |];
//   [| 0;  0;  0;  0;  0;  0;  0;  0;  0;  0;  0;  0;  0;  2 |];
//   [| 0;  0;  0;  0;  0;  0;  0;  0;  0;  0;  0;  0;  0;  0 |];
// |]

// output:
// 12
